pkg/kubectl/cmd/drain/drain.go

```diff
@@ -344,218 +304,43 @@ func (o *DrainOptions) RunDrain() error {
-	return false, &warning{daemonsetWarning}, nil
-}
-
-func mirrorPodFilter(pod corev1.Pod) (bool, *warning, *fatal) {
-	if _, found := pod.ObjectMeta.Annotations[corev1.MirrorPodAnnotationKey]; found {
-		return false, nil, nil
-	}
-	return true, nil, nil
-}
+	if err := o.deleteOrEvictPods(list.Pods()); err != nil {
+		pendingList, newErrs := o.drainer.GetPodsForDeletion(nodeInfo.Name)
 
-func hasLocalStorage(pod corev1.Pod) bool {
-	for _, volume := range pod.Spec.Volumes {
-		if volume.EmptyDir != nil {
-			return true
-		}
-	}
-
-	return false
-}
-
-func (o *DrainOptions) localStorageFilter(pod corev1.Pod) (bool, *warning, *fatal) {
-	if !hasLocalStorage(pod) {
-		return true, nil, nil
-	}
-	// Any finished pod can be removed.
-	if pod.Status.Phase == corev1.PodSucceeded || pod.Status.Phase == corev1.PodFailed {
-		return true, nil, nil
-	}
-	if !o.DeleteLocalData {
-		return false, nil, &fatal{localStorageFatal}
-	}
-
-	return true, &warning{localStorageWarning}, nil
-}
-
-// Map of status message to a list of pod names having that status.
-type podStatuses map[string][]string
-
-func (ps podStatuses) Message() string {
-	msgs := []string{}
-
-	for key, pods := range ps {
-		msgs = append(msgs, fmt.Sprintf("%s: %s", key, strings.Join(pods, ", ")))
-	}
-	return strings.Join(msgs, "; ")
-}
-
-// getPodsForDeletion receives resource info for a node, and returns all the pods from the given node that we
-// are planning on deleting. If there are any pods preventing us from deleting, we return that list in an error.
-func (o *DrainOptions) getPodsForDeletion(nodeInfo *resource.Info) (pods []corev1.Pod, err error) {
-	labelSelector, err := labels.Parse(o.PodSelector)
-	if err != nil {
-		return pods, err
-	}
-
-	podList, err := o.client.CoreV1().Pods(metav1.NamespaceAll).List(metav1.ListOptions{
-		LabelSelector: labelSelector.String(),
-		FieldSelector: fields.SelectorFromSet(fields.Set{"spec.nodeName": nodeInfo.Name}).String()})
-	if err != nil {
-		return pods, err
-	}
-
-	ws := podStatuses{}
-	fs := podStatuses{}
-
-	for _, pod := range podList.Items {
-		podOk := true
-		for _, filt := range []podFilter{o.daemonsetFilter, mirrorPodFilter, o.localStorageFilter, o.unreplicatedFilter} {
-			filterOk, w, f := filt(pod)
-
-			podOk = podOk && filterOk
-			if w != nil {
-				ws[w.string] = append(ws[w.string], pod.Name)
-			}
-			if f != nil {
-				fs[f.string] = append(fs[f.string], pod.Name)
-			}
-
-			// short-circuit as soon as pod not ok
-			// at that point, there is no reason to run pod
-			// through any additional filters
-			if !podOk {
-				break
-			}
+		fmt.Fprintf(o.ErrOut, "There are pending pods in node %q when an error occurred: %v\n", nodeInfo.Name, err)
+		for _, pendingPod := range pendingList.Pods() {
+			fmt.Fprintf(o.ErrOut, "%s/%s\n", "pod", pendingPod.Name)
 		}
-		if podOk {
-			pods = append(pods, pod)
+		if newErrs != nil {
+			fmt.Fprintf(o.ErrOut, "following errors also occurred:\n%s", utilerrors.NewAggregate(newErrs))
```

Comment:
Should newErrs be aggregated into err ?

Reply:
You mean from the library call? That can be discussed. There are still quite a few improvements to be made.

Reply:
What do you think of the following ?
```
diff --git a/pkg/kubectl/cmd/drain/drain.go b/pkg/kubectl/cmd/drain/drain.go
index 8e98e84ee6..76beaef2cf 100644
--- a/pkg/kubectl/cmd/drain/drain.go
+++ b/pkg/kubectl/cmd/drain/drain.go
@@ -322,6 +322,8 @@ func (o *DrainCmdOptions) deleteOrEvictPodsSimple(nodeInfo *resource.Info) error
                }
                if newErrs != nil {
                        fmt.Fprintf(o.ErrOut, "following errors also occurred:\n%s", utilerrors.NewAggregate(newErrs))
+                       newErrs = append(newErrs, err)
+                       err = utilerrors.NewAggregate(newErrs)
                }
                return err
        }
```

Reply:
It could be good, but I am not entirely sure, as the intention here was to avoid drastic changes of kubectl output, or general behaviour. This code originally returned the first error and printed new errors, so I kept it the same way. Not say it has to be that way, please open another PR if you strongly believe this must change. Small PR would be easy to discuss.